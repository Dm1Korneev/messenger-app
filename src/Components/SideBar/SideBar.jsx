import React, { useEffect } from 'react';
import clsx from 'clsx';
import PropTypes from 'prop-types';
import { makeStyles } from '@material-ui/core/styles';
import IconButton from '@material-ui/core/IconButton';
import Drawer from '@material-ui/core/Drawer';
import ListItem from '@material-ui/core/ListItem';
import List from '@material-ui/core/List';
import ListItemText from '@material-ui/core/ListItemText';
import Divider from '@material-ui/core/Divider';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ChevronLeftIcon from '@material-ui/icons/ChevronLeft';
import AddBox from '@material-ui/icons/AddBox';

import { DRAWER_WIDTH, RELOAD_PERIOD } from 'Constants';
import Chat from 'Components/Chat';

const useStyles = makeStyles((theme) => ({
  drawer: {
    overflowX: 'hidden',
    whiteSpace: 'nowrap',
    position: 'relative',
  },
  drawerOpen: {
    width: DRAWER_WIDTH,
    transition: theme.transitions.create('width', {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.enteringScreen,
    }),
  },
  drawerClose: {
    width: theme.spacing(9),
    transition: theme.transitions.create('width', {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.leavingScreen,
    }),
  },
  toolbarIcon: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'flex-end',
    padding: theme.spacing(0, 1),
    ...theme.mixins.toolbar,
  },
  addChatIcon: {
    fontSize: theme.spacing(6),
    marginLeft: -theme.spacing(0.5),
  },
}));

const SideBar = ({
  getChats, drawerIsOpen,
  chats,
  activeChat,
  onDrawerClose,
  changeActiveChat,
  openAddChatDialog,
  openModifyChatDialog,
}) => {
  useEffect(() => {
    const interval = setInterval(getChats, RELOAD_PERIOD);
    return () => {
      if (interval) {
        clearInterval(interval);
      }
    };
  }, [getChats]);
  const classes = useStyles();

  return (
    <Drawer
      variant="permanent"
      classes={{
        paper: clsx(
          classes.drawer, {
            [classes.drawerOpen]: drawerIsOpen,
            [classes.drawerClose]: !drawerIsOpen,
          },
        ),
      }}
      open={drawerIsOpen}
    >
      <div className={classes.toolbarIcon}>
        <IconButton onClick={onDrawerClose}>
          <ChevronLeftIcon />
        </IconButton>
      </div>
      <Divider />
      <List>
        <ListItem button key="add_chat" onClick={openAddChatDialog}>
          <ListItemIcon>
            <AddBox color="primary" className={classes.addChatIcon} />
          </ListItemIcon>
          <ListItemText primary="Add chat" />
        </ListItem>
        {chats.map((chat) => {
          const { _id } = chat;
          return (
            <Chat
              chat={chat}
              key={_id}
              selected={_id === activeChat}
              chatOnClick={() => changeActiveChat({ activeChat: _id })}
              chatModifyOnClick={openModifyChatDialog}
            />
          );
        })}
      </List>
    </Drawer>
  );
};

SideBar.defaultProps = {
  activeChat: null,
  chats: [],
};
SideBar.propTypes = {
  getChats: PropTypes.func.isRequired,
  drawerIsOpen: PropTypes.bool.isRequired,
  chats: PropTypes.arrayOf(PropTypes.instanceOf(Object)),
  activeChat: PropTypes.string,
  onDrawerClose: PropTypes.func.isRequired,
  changeActiveChat: PropTypes.func.isRequired,
  openAddChatDialog: PropTypes.func.isRequired,
  openModifyChatDialog: PropTypes.func.isRequired,
};

export default SideBar;

