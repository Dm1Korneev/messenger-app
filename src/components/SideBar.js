import React, { Component } from "react";
import withStyles from "@material-ui/core/styles/withStyles";
import IconButton from "@material-ui/core/IconButton";
import classNames from "classnames";
import Drawer from "@material-ui/core/Drawer";
import ListItem from "@material-ui/core/ListItem";
import List from "@material-ui/core/List";
import ListItemText from "@material-ui/core/ListItemText";
import ChevronLeftIcon from "@material-ui/icons/ChevronLeft";
import Divider from "@material-ui/core/Divider";
import { DRAWER_WIDTH } from "../common/constants";
import AddBoxIcon from "@material-ui/icons/AddBox";
import Chat from "./Chat";

import { RELOAD_PERIOD } from "../common/constants";

class SideBar extends Component {
  constructor(props) {
    super(props);

    this.reloadChatsList = props.reloadChatsList;
    this.interval = undefined;
  }

  componentDidMount() {
    this.interval = setInterval(this.reloadChatsList, RELOAD_PERIOD);
  }

  componentWillUnmount() {
    if (this.interval) {
      clearInterval(this.interval);
    }
  }

  render() {
    const {
      drawerIsOpen,
      chats,
      activeChat,
      onDrawerClose,
      classes,
      changeActiveChat,
      openAddChatDialog,
      openModifyChatDialog
    } = this.props;

    return (
      <Drawer
        variant="permanent"
        classes={{
          paper: classNames(
            classes.drawerPaper,
            !drawerIsOpen && classes.drawerPaperClose
          )
        }}
        open={drawerIsOpen}
      >
        <div className={classes.toolbarIcon}>
          <IconButton onClick={onDrawerClose}>
            <ChevronLeftIcon />
          </IconButton>
        </div>
        <Divider />
        <List>
          <ListItem button key={"add_chat"} onClick={openAddChatDialog}>
            <AddBoxIcon color="primary" className={classes.addChatIcon} />
            <ListItemText primary={"Add chat"} />
          </ListItem>

          {chats.allIds.map(value => {
            const chat = chats.byId[value];
            const { _id } = chat;
            return (
              <Chat
                chat={chat}
                key={_id}
                selected={_id === activeChat}
                chatOnClick={changeActiveChat}
                chatModifyOnClick={openModifyChatDialog}
              />
            );
          })}
        </List>
      </Drawer>
    );
  }
}

const styles = theme => ({
  drawerPaper: {
    position: "relative",
    whiteSpace: "nowrap",
    width: DRAWER_WIDTH,
    transition: theme.transitions.create("width", {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.enteringScreen
    })
  },
  drawerPaperClose: {
    overflowX: "hidden",
    transition: theme.transitions.create("width", {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.leavingScreen
    }),
    width: theme.spacing.unit * 7,
    [theme.breakpoints.up("sm")]: {
      width: theme.spacing.unit * 9
    }
  },
  toolbarIcon: {
    display: "flex",
    alignItems: "center",
    justifyContent: "flex-end",
    padding: "0 8px",
    ...theme.mixins.toolbar
  },
  addChatIcon: {
    fontSize: theme.spacing.unit * 7,
    marginLeft: -theme.spacing.unit
  }
});

export default withStyles(styles)(SideBar);
