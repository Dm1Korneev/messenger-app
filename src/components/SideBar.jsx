import React from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';

import withStyles from '@material-ui/core/styles/withStyles';
import IconButton from '@material-ui/core/IconButton';
import Drawer from '@material-ui/core/Drawer';
import ListItem from '@material-ui/core/ListItem';
import List from '@material-ui/core/List';
import ListItemText from '@material-ui/core/ListItemText';
import ChevronLeftIcon from '@material-ui/icons/ChevronLeft';
import Divider from '@material-ui/core/Divider';
import AddBoxIcon from '@material-ui/icons/AddBox';

import { DRAWER_WIDTH, RELOAD_PERIOD } from 'Common/constants';

import Chat from './Chat';

class SideBar extends React.Component {
  constructor(props) {
    super(props);

    this.getChats = props.getChats;
    this.interval = undefined;
  }

  componentDidMount() {
    this.interval = setInterval(this.getChats, RELOAD_PERIOD);
  }

  componentWillUnmount() {
    if (this.interval) {
      clearInterval(this.interval);
    }
  }

  render() {
    const {
      drawerIsOpen,
      chats,
      activeChat,
      onDrawerClose,
      classes,
      changeActiveChat,
      openAddChatDialog,
      openModifyChatDialog,
    } = this.props;

    return (
      <Drawer
        variant="permanent"
        classes={{
          paper: classNames(
            classes.drawerPaper,
            !drawerIsOpen && classes.drawerPaperClose,
          ),
        }}
        open={drawerIsOpen}
      >
        <div className={classes.toolbarIcon}>
          <IconButton onClick={onDrawerClose}>
            <ChevronLeftIcon />
          </IconButton>
        </div>
        <Divider />
        <List>
          <ListItem button key="add_chat" onClick={openAddChatDialog}>
            <AddBoxIcon color="primary" className={classes.addChatIcon} />
            <ListItemText primary="Add chat" />
          </ListItem>

          {chats.allIds.map((value) => {
            const chat = chats.byId[value];
            const { _id } = chat;
            return (
              <Chat
                chat={chat}
                key={_id}
                selected={_id === activeChat}
                chatOnClick={changeActiveChat}
                chatModifyOnClick={openModifyChatDialog}
              />
            );
          })}
        </List>
      </Drawer>
    );
  }
}

SideBar.defaultProps = {
  activeChat: null,
  chats: [],
};
SideBar.propTypes = {
  classes: PropTypes.instanceOf(Object).isRequired,
  getChats: PropTypes.func.isRequired,
  drawerIsOpen: PropTypes.bool.isRequired,
  chats: PropTypes.shape({
    allIds: PropTypes.arrayOf(PropTypes.string),
    byId: PropTypes.object,
  }),
  activeChat: PropTypes.string,
  onDrawerClose: PropTypes.func.isRequired,
  changeActiveChat: PropTypes.func.isRequired,
  openAddChatDialog: PropTypes.func.isRequired,
  openModifyChatDialog: PropTypes.func.isRequired,
};

const styles = (theme) => ({
  drawerPaper: {
    position: 'relative',
    whiteSpace: 'nowrap',
    width: DRAWER_WIDTH,
    transition: theme.transitions.create('width', {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.enteringScreen,
    }),
  },
  drawerPaperClose: {
    overflowX: 'hidden',
    transition: theme.transitions.create('width', {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.leavingScreen,
    }),
    width: theme.spacing.unit * 7,
    [theme.breakpoints.up('sm')]: {
      width: theme.spacing.unit * 9,
    },
  },
  toolbarIcon: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'flex-end',
    padding: '0 8px',
    ...theme.mixins.toolbar,
  },
  addChatIcon: {
    fontSize: theme.spacing.unit * 7,
    marginLeft: -theme.spacing.unit,
  },
});

export default withStyles(styles)(SideBar);
