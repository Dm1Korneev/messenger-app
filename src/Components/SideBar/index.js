import commonHoc from 'Components/commonHoc';

import {
  changeActiveChat,
  getChats,
  openModifyChatDialog,
  setAddChatDialogIsOpen,
  setDrawerIsOpen,
} from 'Redux/actions';

import { chatsArraySelector } from 'Selectors/chats';
import { activeChatIdSelector, drawerIsOpenSelector } from 'Selectors/session';

import SideBar from './SideBar';

const mapStateToProps = (state) => ({
  chats: chatsArraySelector(state),
  activeChat: activeChatIdSelector(state),
  drawerIsOpen: drawerIsOpenSelector(state),
});

const mapDispatchToProps = {
  onDrawerClose: () => setDrawerIsOpen(false),
  changeActiveChat,
  openAddChatDialog: () => setAddChatDialogIsOpen(true),
  openModifyChatDialog,
  getChats,
};

const Container = commonHoc(SideBar, {
  mapStateToProps,
  mapDispatchToProps,
});

export { Container as SideBar };
