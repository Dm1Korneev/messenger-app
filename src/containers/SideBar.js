import { connect } from "react-redux";
import {
  setDrawerIsOpen,
  changeActiveChat,
  setAddChatDialogIsOpen,
  openModifyChatDialog,
  reloadChatsList
} from "../redux/actions";
import SideBar from "../components/SideBar";

const mapStateToProps = state => ({
  chats: state.chats,
  activeChat: state.session.activeChat,
  drawerIsOpen: state.session.drawerIsOpen
});

const mapDispatchToProps = dispatch => ({
  onDrawerClose: () => dispatch(setDrawerIsOpen(false)),
  changeActiveChat: activeChat => dispatch(changeActiveChat(activeChat)),
  openAddChatDialog: () => dispatch(setAddChatDialogIsOpen(true)),
  openModifyChatDialog: chat => dispatch(openModifyChatDialog(chat)),
  reloadChatsList: () => dispatch(reloadChatsList())
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(SideBar);
