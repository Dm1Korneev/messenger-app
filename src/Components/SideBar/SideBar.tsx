import AddBox from '@mui/icons-material/AddBox';
import ChevronLeftIcon from '@mui/icons-material/ChevronLeft';
import Box from '@mui/material/Box';
import Divider from '@mui/material/Divider';
import Drawer from '@mui/material/Drawer';
import IconButton from '@mui/material/IconButton';
import List from '@mui/material/List';
import ListItem from '@mui/material/ListItem';
import ListItemIcon from '@mui/material/ListItemIcon';
import ListItemText from '@mui/material/ListItemText';

import { Chat } from 'Components/Chat';
import { DRAWER_WIDTH } from 'Constants';
import { useChats } from 'Hooks';

type SideBarProps = {
  chatModifyOnClick: (id: string) => void
  onChatAddClick: () => void
  isDrawerOpen: boolean
  onDrawerClose: () => void
  activeChatId?: string
  onChatClick: (chatId: string) => void
}

export const SideBar = ({
  chatModifyOnClick, onChatAddClick, isDrawerOpen, onDrawerClose, activeChatId, onChatClick,
}: SideBarProps) => {
  const { data: chats } = useChats();

  return (
    <Drawer
      open={isDrawerOpen}
      sx={(theme) => {
        const additionalStyles = isDrawerOpen ? {
          width: DRAWER_WIDTH,
          transition: theme.transitions.create('width', {
            easing: theme.transitions.easing.sharp,
            duration: theme.transitions.duration.enteringScreen,
          }),
        } : {
          width: theme.spacing(9),
          transition: theme.transitions.create('width', {
            easing: theme.transitions.easing.sharp,
            duration: theme.transitions.duration.leavingScreen,
          }),
        };

        return {
          '.MuiDrawer-paper': {
            overflowX: 'hidden',
            whiteSpace: 'nowrap',
            position: 'relative',
            ...additionalStyles,
          },
        };
      }}
      variant="permanent"
    >
      <Box sx={(theme) => ({
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'flex-end',
        p: theme.spacing(0, 1),
        ...theme.mixins.toolbar,
      })}
      >
        <IconButton onClick={onDrawerClose} size="large">
          <ChevronLeftIcon />
        </IconButton>
      </Box>
      <Divider />
      <List>
        <ListItem key="add_chat" button onClick={onChatAddClick}>
          <ListItemIcon>
            <AddBox
              color="primary"
              sx={(theme) => ({
                fontSize: theme.spacing(6),
                ml: -0.5,
              })}
            />
          </ListItemIcon>
          <ListItemText primary="Add chat" />
        </ListItem>
        {chats.map((chat) => {
          const { _id } = chat;
          return (
            <Chat
              key={_id}
              chat={chat}
              chatModifyOnClick={chatModifyOnClick}
              chatOnClick={() => onChatClick(_id)}
              selected={_id === activeChatId}
            />
          );
        })}
      </List>
    </Drawer>
  );
};
