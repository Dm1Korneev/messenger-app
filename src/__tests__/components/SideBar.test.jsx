import React from 'react';

import SideBar from 'Components/SideBar';

jest.mock('Components/Chat', () => global.mockComponent('Chat'));
jest.mock('Components/UserNameField', () => global.mockComponent('UserNameField'));
jest.mock('Components/UserPasswordField', () => global.mockComponent('UserPasswordField'));
jest.mock('Components/AvatarSelector', () => global.mockComponent('AvatarSelector'));

jest.mock('@material-ui/core/IconButton', () => global.mockComponent('IconButton'));
jest.mock('@material-ui/core/Drawer', () => global.mockComponent('Drawer'));
jest.mock('@material-ui/core/ListItem', () => global.mockComponent('ListItem'));
jest.mock('@material-ui/core/List', () => global.mockComponent('List'));
jest.mock('@material-ui/core/ListItemText', () => global.mockComponent('ListItemText'));
jest.mock('@material-ui/core/Divider', () => global.mockComponent('Divider'));

jest.mock('@material-ui/icons/ChevronLeft', () => global.mockComponent('ChevronLeft'));
jest.mock('@material-ui/icons/AddBox', () => global.mockComponent('AddBox'));

const props = {
  getChats: jest.fn(),
  drawerIsOpen: false,
  onDrawerClose: jest.fn(),
  changeActiveChat: jest.fn(),
  openAddChatDialog: jest.fn(),
  openModifyChatDialog: jest.fn(),
  onSignIn: jest.fn(),
  onRegister: jest.fn(),
};

describe('render SideBar component', () => {
  let wrapper;

  beforeAll(() => {
    wrapper = global.mount(
      <SideBar {...props} />,
    );
  });

  test('component is render', () => {
    expect(wrapper.find('SideBar').length).toBe(1);
  });

  test('Drawer subcomponent is render', () => {
    wrapper.update();
    expect(wrapper.find({ originalcomponent: 'Drawer' }).length).toBe(1);
  });

  test('ChevronLeft subcomponent is render', () => {
    wrapper.update();
    expect(wrapper.find({ originalcomponent: 'ChevronLeft' }).length).toBe(1);
  });
});

describe('snapshot-test SideBar component', () => {
  test('Renders correct properties', () => {
    global.mountExpect(
      <SideBar {...props} />,
    ).toMatchSnapshot();
  });
});
