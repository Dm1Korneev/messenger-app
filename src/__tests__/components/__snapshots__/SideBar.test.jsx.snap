// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot-test SideBar component Renders correct properties 1`] = `
<WithStyles(SideBar)
  changeActiveChat={[MockFunction]}
  drawerIsOpen={false}
  getChats={[MockFunction]}
  onDrawerClose={[MockFunction]}
  onRegister={[MockFunction]}
  onSignIn={[MockFunction]}
  openAddChatDialog={[MockFunction]}
  openModifyChatDialog={[MockFunction]}
>
  <SideBar
    activeChat={null}
    changeActiveChat={[MockFunction]}
    chats={Array []}
    classes={
      Object {
        "addChatIcon": "SideBar-addChatIcon-4",
        "drawerPaper": "SideBar-drawerPaper-1",
        "drawerPaperClose": "SideBar-drawerPaperClose-2",
        "toolbarIcon": "SideBar-toolbarIcon-3",
      }
    }
    drawerIsOpen={false}
    getChats={[MockFunction]}
    onDrawerClose={[MockFunction]}
    onRegister={[MockFunction]}
    onSignIn={[MockFunction]}
    openAddChatDialog={[MockFunction]}
    openModifyChatDialog={[MockFunction]}
  >
    <Component
      classes={
        Object {
          "paper": "SideBar-drawerPaper-1 SideBar-drawerPaperClose-2",
        }
      }
      open={false}
      variant="permanent"
    >
      <div
        originalcomponent="Drawer"
        testprops={
          Object {
            "children": Array [
              <div
                className="SideBar-toolbarIcon-3"
              >
                <Unknown
                  onClick={[MockFunction]}
                >
                  <Unknown />
                </Unknown>
              </div>,
              <Unknown />,
              <Unknown>
                <Unknown
                  button={true}
                  onClick={[MockFunction]}
                >
                  <Unknown
                    className="SideBar-addChatIcon-4"
                    color="primary"
                  />
                  <Unknown
                    primary="Add chat"
                  />
                </Unknown>
              </Unknown>,
            ],
            "classes": Object {
              "paper": "SideBar-drawerPaper-1 SideBar-drawerPaperClose-2",
            },
            "open": false,
            "variant": "permanent",
          }
        }
      >
        <div
          className="SideBar-toolbarIcon-3"
        >
          <Component
            onClick={[MockFunction]}
          >
            <div
              originalcomponent="IconButton"
              testprops={
                Object {
                  "children": <Unknown />,
                  "onClick": [MockFunction],
                }
              }
            >
              <Component>
                <div
                  originalcomponent="ChevronLeft"
                  testprops={Object {}}
                />
              </Component>
            </div>
          </Component>
        </div>
        <Component>
          <div
            originalcomponent="Divider"
            testprops={Object {}}
          />
        </Component>
        <Component>
          <div
            originalcomponent="List"
            testprops={
              Object {
                "children": Array [
                  <Unknown
                    button={true}
                    onClick={[MockFunction]}
                  >
                    <Unknown
                      className="SideBar-addChatIcon-4"
                      color="primary"
                    />
                    <Unknown
                      primary="Add chat"
                    />
                  </Unknown>,
                  Array [],
                ],
              }
            }
          >
            <Component
              button={true}
              key="add_chat"
              onClick={[MockFunction]}
            >
              <div
                originalcomponent="ListItem"
                testprops={
                  Object {
                    "button": true,
                    "children": Array [
                      <Unknown
                        className="SideBar-addChatIcon-4"
                        color="primary"
                      />,
                      <Unknown
                        primary="Add chat"
                      />,
                    ],
                    "onClick": [MockFunction],
                  }
                }
              >
                <Component
                  className="SideBar-addChatIcon-4"
                  color="primary"
                >
                  <div
                    originalcomponent="AddBox"
                    testprops={
                      Object {
                        "className": "SideBar-addChatIcon-4",
                        "color": "primary",
                      }
                    }
                  />
                </Component>
                <Component
                  primary="Add chat"
                >
                  <div
                    originalcomponent="ListItemText"
                    testprops={
                      Object {
                        "primary": "Add chat",
                      }
                    }
                  />
                </Component>
              </div>
            </Component>
          </div>
        </Component>
      </div>
    </Component>
  </SideBar>
</WithStyles(SideBar)>
`;
